gosax.go

```diff
@@ -161,6 +162,66 @@ func ParseFile(filename string, cb SaxCallbacks) error {
 	return nil
 }
 
+// A better SAX parsing routine. parse an XML in-memory buffer, with cb as
+// the callbacks.
+func ParseMem(buf bytes.Buffer, cb SaxCallbacks) error {
+	//var cfilename *C.char = C.CString(filename)
+	//defer C.free(unsafe.Pointer(cfilename))
+
+	// newHandlerStruct zeroes out all the pointers; we assign only those that
+	// are passed as non-nil in SaxCallbacks.
+	SAXhandler := C.newHandlerStruct()
+
+	if cb.StartDocument != nil {
+		SAXhandler.startDocument = C.startDocumentSAXFunc(C.startDocumentCgo)
+	}
+
+	if cb.EndDocument != nil {
+		SAXhandler.endDocument = C.endDocumentSAXFunc(C.endDocumentCgo)
+	}
+
+	if cb.StartElement != nil {
+		SAXhandler.startElement = C.startElementSAXFunc(C.startElementCgo)
+	}
+	// StartElementNoAttr overrides StartElement
+	if cb.StartElementNoAttr != nil {
+		SAXhandler.startElement = C.startElementSAXFunc(C.startElementNoAttrCgo)
+	}
+
+	if cb.EndElement != nil {
+		SAXhandler.endElement = C.endElementSAXFunc(C.endElementCgo)
+	}
+
+	if cb.Characters != nil {
+		SAXhandler.characters = C.charactersSAXFunc(C.charactersCgo)
+	}
+	// CharactersRaw overrides Characters
+	if cb.CharactersRaw != nil {
+		SAXhandler.characters = C.charactersSAXFunc(C.charactersRawCgo)
+	}
+
+	// Pack the callbacks structure into an opaque unsafe.Pointer which we'll
+	// pass to C as user_data, and C will pass it back to our Go callbacks.
+	user_data := pointer.Save(&cb)
+	defer pointer.Unref(user_data)
+
+	bufPointer := C.malloc(C.size_t(buf.Len()))
```

Comment:
Another request: please add a comment here explaining what's going on with the buffers